### PR TITLE
Feat/met 1826 backend support for new executions list

### PR DESCRIPTION
**Pull request by jochen-vermeulen**

### CODE REVIEW

File: metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java [hunk 1]
```diff
@@ -346,6 +354,122 @@ private AbstractMetisPlugin getFirstOrLastFinishedWorkflowExecutionPluginByDatas
...
+
+    // Step 6: Join with the dataset and the execution
+    final String datasetCollectionName = morphiaDatastoreProvider.getDatastore()
+        .getCollection(Dataset.class).getName();
+    final String executionCollectionName = morphiaDatastoreProvider.getDatastore()
+        .getCollection(WorkflowExecution.class).getName();
+    final String datasetListField = "datasetList";
+    final String executionListField = "executionList";
+    pipeline.lookup(datasetCollectionName, DATASET_ID, DatasetDao.DATASET_ID, datasetListField);
+    pipeline.lookup(executionCollectionName, "_id", "_id", executionListField);
```
Comment (stzanakis) [inline metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java]: Why is this join required?
Is it that the first lookup(Join) will produce the workflowExecution with the addition of the "datasetList" which in fact can only contain one item, the one that comes from the dataset collection?

Comment (jochen-vermeulen) [reply metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java]: As discussed: we don't have the execution here: by repeatedly projecting we lose all the fields we don't explicitly name. This is better than naming all those fields.

File: metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java [hunk 2]
```diff
@@ -346,6 +354,122 @@ private AbstractMetisPlugin getFirstOrLastFinishedWorkflowExecutionPluginByDatas
...
+        .limit(getWorkflowExecutionsPerRequest());
+
+    // Step 6: Join with the dataset and the execution
+    final String datasetCollectionName = morphiaDatastoreProvider.getDatastore()
+        .getCollection(Dataset.class).getName();
+    final String executionCollectionName = morphiaDatastoreProvider.getDatastore()
+        .getCollection(WorkflowExecution.class).getName();
+    final String datasetListField = "datasetList";
+    final String executionListField = "executionList";
+    pipeline.lookup(datasetCollectionName, DATASET_ID, DatasetDao.DATASET_ID, datasetListField);
```
[the remark this answers is not in the document]
Comment (jochen-vermeulen) [reply metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java]: I think I don't agree with this suggestion. The DATASET_ID in WorkflowExecutionDao should contain the name of the field in the WorkflowExecution object. The one in DatasetDao contains the name of the field in the Dataset object. They are the same now, but maybe one day that changes? I specifically mean the one in the Dataset object here.

File: metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java [hunk 3]
```diff
@@ -346,6 +354,122 @@ private AbstractMetisPlugin getFirstOrLastFinishedWorkflowExecutionPluginByDatas
...
+    }
+
+    // Step 2: Add difference integers on whether the status is INQUEUE or RUNNING.
+    final String statusInQueueField = "statusInQueue";
+    final String statusRunningField = "statusRunning";
+    final int inqueueDifferennce = DEFAULT_POSITION_IN_OVERVIEW - INQUEUE_POSITION_IN_OVERVIEW;
+    final int runningDifferennce = DEFAULT_POSITION_IN_OVERVIEW - RUNNING_POSITION_IN_OVERVIEW;
+    pipeline.project(
+        Projection.projection(statusInQueueField, Projection.expression("$cond",
+            Projection.expression(FilterOperator.EQUAL.val(), "INQUEUE", "$" + WORKFLOW_STATUS),
```
[the remark this answers is not in the document]
Comment (jochen-vermeulen) [reply metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java]: Fixed.

File: metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/rest/execution/overview/PluginProgress.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package eu.europeana.metis.core.rest.execution.overview;
+
+import eu.europeana.metis.core.workflow.plugins.ExecutionProgress;
+
+/**
+ * This class contains executionProgress information on a plugin's execution.
+ */
+public class PluginProgress {
```
[the remark this answers is not in the document]
Comment (jochen-vermeulen) [reply metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/rest/execution/overview/PluginProgress.java]: Two reasons. 1. I believe that eventually it is better to have the database objects and the UI objects separated. This means that we can easily make changes to them without accidentally breaking something else. 2. Once all these objects are no longer the same as their database versions, it feels somewhat inconsistent to have this one still the same. And of course 3. the overhead! :)

File: metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/rest/execution/overview/ExecutionProgress.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import eu.europeana.metis.core.workflow.plugins.AbstractMetisPlugin;
+import eu.europeana.metis.core.workflow.plugins.PluginStatus;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This object contains all information regarding the workflow execution's progress.
+ */
+public class ExecutionProgress {
```
[the remark this answers is not in the document]
Comment (jochen-vermeulen) [reply metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/rest/execution/overview/ExecutionProgress.java]: Fixed.

File: metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java [hunk 6]
```diff
@@ -346,6 +354,122 @@ private AbstractMetisPlugin getFirstOrLastFinishedWorkflowExecutionPluginByDatas
...
+      final Query<WorkflowExecution> query =
+          morphiaDatastoreProvider.getDatastore().createQuery(WorkflowExecution.class);
+      query.field(DATASET_ID).in(datasetIds);
+      pipeline.match(query);
+    }
+
+    // Step 2: Add difference integers on whether the status is INQUEUE or RUNNING.
+    final String statusInQueueField = "statusInQueue";
+    final String statusRunningField = "statusRunning";
+    final int inqueueDifferennce = DEFAULT_POSITION_IN_OVERVIEW - INQUEUE_POSITION_IN_OVERVIEW;
```
[the remark this answers is not in the document]
Comment (jochen-vermeulen) [reply metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java]: Ok, I fixed it.

File: metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/rest/execution/overview/PluginProgress.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/rest/execution/overview/PluginProgress.java]: Ok I see your point and makes sense if we start doing it everywhere. But if we agree to start having different objects for the frontend than the ones for the database, perhaps we should start having naming conventions, like adding a suffix "View" in the class name or something similar.

File: metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (stzanakis) [reply metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java]: Indeed I agree

File: metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java (hunk 1 above)
Comment (stzanakis) [reply metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowExecutionDao.java]: Yes

File: metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/rest/execution/overview/PluginProgress.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jochen-vermeulen) [reply metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/rest/execution/overview/PluginProgress.java]: OK. I fixed it.

**Pull request merged**